### PR TITLE
Fix validation for user-defined formats

### PR DESCRIPTION
**Pull request by analogue**
See #48 

This farms out user-defined format validation to `jsonschema`.

### CODE REVIEW

File: bravado_core/formatter.py [hunk 1]
```diff
@@ -88,6 +110,50 @@ class SwaggerFormat(namedtuple('SwaggerFormat',
...
+    Build and cache a :class:`jsonschema.FormatChecker` for validating
+    user-defined Swagger formats.
+
+    :rtype: :class:`jsonschema.FormatChecker`
+    """
+    global _format_checker
+    if _format_checker is None:
+        _format_checker = FormatChecker()
+        for swagger_format in _user_defined_formats:
+            _format_checker.checkers[swagger_format.format] = (
```
Comment (prat0318) [inline bravado_core/formatter.py]: Instead use [the API](http://python-jsonschema.readthedocs.org/en/latest/validate/#jsonschema.FormatChecker.checkers) available for adding a checker?

File: bravado_core/formatter.py [hunk 2]
```diff
@@ -88,6 +110,50 @@ class SwaggerFormat(namedtuple('SwaggerFormat',
...
+
+def get_format_checker():
+    """
+    Build and cache a :class:`jsonschema.FormatChecker` for validating
+    user-defined Swagger formats.
+
+    :rtype: :class:`jsonschema.FormatChecker`
+    """
+    global _format_checker
+    if _format_checker is None:
```
Comment (prat0318) [inline bravado_core/formatter.py]: Won't global format_checker increase the pain described in #50?

File: bravado_core/formatter.py [hunk 3]
```diff
@@ -88,6 +110,50 @@ class SwaggerFormat(namedtuple('SwaggerFormat',
     """
 
 
+def return_true_wrapper(validate_func):
+    """Decorator for the SwaggerFormat.validate function to always return True.
+
+    The contract for `SwaggerFormat.validate` is to raise an exception
+    when validation fails. However, the contract for jsonschema's
+    validate function is to raise an exception or return True. This wrapper
+    bolts-on the `return True` part.
```
Comment (prat0318) [inline bravado_core/formatter.py]: Also link it to the jsonschema doc : http://python-jsonschema.readthedocs.org/en/latest/validate/#jsonschema.FormatChecker

File: bravado_core/formatter.py [hunk 4]
```diff
@@ -46,18 +48,38 @@ def to_python(spec, value):
...
     """
     global _formatters
     _formatters[swagger_format.format] = swagger_format
 
+    # Need to maintain a separate list of UDFs for jsonschema validation
+    global _user_defined_formats
+    _user_defined_formats.append(swagger_format)
+
+
+def unregister_format(swagger_format):
```
Comment (prat0318) [inline bravado_core/formatter.py]: When/why will an application call this function?

File: docs/source/changelog.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+../../CHANGELOG.rst
```
Comment (prat0318) [inline docs/source/changelog.rst]: does this work?

File: tests/validate/validate_array_test.py [hunk 6]
```diff
@@ -54,3 +55,45 @@ def test_uniqueItems_false(int_array_spec):
...
+        'type': 'array',
+        'items': {
+            'type': 'string',
+            'format': 'ipv4',
+        }
+    }
+    request_body = ['not_an_ip_address']
+    with pytest.raises(ValidationError) as excinfo:
+        validate_array(ipaddress_array_spec, request_body)
+    assert "'not_an_ip_address' is not a 'ipv4'" in str(excinfo.value)
```
Comment (prat0318) [inline tests/validate/validate_array_test.py]: but where is ipv4 format registered?

File: bravado_core/formatter.py (hunk 1 above)
Comment (analogue) [inline bravado_core/formatter.py]: The API only provides a function decorator.

File: bravado_core/formatter.py (hunk 2 above)
Comment (analogue) [inline bravado_core/formatter.py]: Yes. Better to fix this bug in a backwards compatible way before releasing a non-backward compatible version.

File: bravado_core/formatter.py (hunk 3 above)
Comment (analogue) [inline bravado_core/formatter.py]: Don't really think this is necessary. FormatChecker is integral to the validation of user-defined formats. Any developer looking for more info is going to easily be able to navigate to the source. Furthermore, this isn't a part of the public API.

File: bravado_core/formatter.py (hunk 4 above)
Comment (analogue) [inline bravado_core/formatter.py]: See unit tests

File: docs/source/changelog.rst (hunk 5 above)
Comment (analogue) [inline docs/source/changelog.rst]: Yes

File: tests/validate/validate_array_test.py (hunk 6 above)
Comment (analogue) [inline tests/validate/validate_array_test.py]: As the test method name implies, it is a builtin format.

File: bravado_core/formatter.py (hunk 1 above)
Comment (prat0318) [inline bravado_core/formatter.py]:
```python
validator = return_true_wrapper(swagger_format.validate)
_format_checker.checks(
    swagger_format.format, raises=(SwaggerValidationError,))(validator)
```
This way there is no need to modify `checkers` attribute directly.

File: bravado_core/formatter.py (hunk 2 above)
Comment (prat0318) [inline bravado_core/formatter.py]: I am thinking if there is an alternate way than introducing global _format_checker and _user_defined_formats.

File: bravado_core/formatter.py (hunk 3 above)
Comment (prat0318) [inline bravado_core/formatter.py]: fine with me.

File: bravado_core/formatter.py (hunk 4 above)
Comment (prat0318) [inline bravado_core/formatter.py]: My question was from the point of view of actual use-case. Pyramid registers user-formats during init of pyramid_swagger, but it need not unregister them at any time. Is it meaningful from client side use-case?

File: docs/source/changelog.rst (hunk 5 above)
Comment (prat0318) [inline docs/source/changelog.rst]: cool.

File: bravado_core/formatter.py (hunk 2 above)
Comment (analogue) [inline bravado_core/formatter.py]: What would that way be?

File: bravado_core/formatter.py (hunk 4 above)
Comment (analogue) [inline bravado_core/formatter.py]: I'd say no. Moving this function to the unit test might be an improvement, but then the scope of the function is unnecessarily limited. What is the critical issue here?

File: bravado_core/formatter.py (hunk 1 above)
Comment (analogue) [inline bravado_core/formatter.py]: Ah, the example makes it clear.

File: bravado_core/formatter.py (hunk 4 above)
Comment (prat0318) [inline bravado_core/formatter.py]: We can then get rid of _user_defined_formats. Addition to _format_checker can happen directly with register_format (and now we do not care about rebuilding it).

File: bravado_core/formatter.py (hunk 2 above)
Comment (prat0318) [inline bravado_core/formatter.py]: i can only think of a way to get rid of _udf (described above). i guess we have to live with _format_checker till _formatters and _format_checker are associated with Spec (as described in #50)

File: tests/validate/validate_array_test.py (hunk 6 above)
Comment (prat0318) [inline tests/validate/validate_array_test.py]: oh there are some jsonschema built-ins. they look useful and the user need not register the format with definitions for these to work.

**Pull request merged**